### PR TITLE
repair partition selection

### PR DESCRIPTION
**Pull request by moarychan**
[Query]
[Spring - can we specify partition key when producing events to event hubs using azure-spring-cloud-stream-binder-eventhubs?](https://github.com/Azure/azure-sdk-for-java/issues/19683)

[Currently]
If we set the spring cloud stream configuration `partitionKeyExpression`  or  `partitionKeyExtractorName`, it does not work, the keyExpressionString will be always "'partitionKey-' + headers['scst_partition']".

[Change]
- Support spring cloud stream `partitionKeyExpression` configuration.
  Support spring cloud stream default partition key definition,  `partitionKeyExpression`  or  `partitionKeyExtractorName`, remove the default key partition expression when `producerProperties.isPartitioned()` returns true.

- Update `PartitionSupplier` parameter for EventHubs SDK, get partitionId first, then partitionKey, choose one.
  Event Hubs SDK limits only one value should be setting, `partitionId` or `partitionKey`.

  Construct the `PartitionSupplier` parameter if any valid value is obtained, below are `partitionId` and `partitionKey` value getting priority:
  1.  `partitionId` from message header `azure_partition_id`
  2.  `partitionId` from SpEL expression `headers['scst_partition']`
  3.  `partitionKey` from message header `azure_partition_key`
  4.  `partitionKey` from function expression `Message.getPayload().hashCode()`

### CODE REVIEW

**Comment by stliu** [comment]
would u please elaborate the purpose of this PR more

File: sdk/spring/azure-spring-cloud-stream-binder-eventhubs/src/main/java/com/azure/spring/eventhub/stream/binder/EventHubMessageChannelBinder.java [hunk 1]
```diff
@@ -56,8 +56,12 @@ protected MessageHandler createProducerMessageHandler(ProducerDestination destin
         handler.setSendTimeout(producerProperties.getExtension().getSendTimeout());
         handler.setSendFailureChannel(errorChannel);
         if (producerProperties.isPartitioned()) {
-            handler.setPartitionKeyExpressionString(
+            if (null != producerProperties.getPartitionKeyExpression()) {
```
Comment (stliu) [inline sdk/spring/azure-spring-cloud-stream-binder-eventhubs/src/main/java/com/azure/spring/eventhub/stream/binder/EventHubMessageChannelBinder.java:59]: personally I would not like this expression, put this *null* first, better to use more common/straight way to check

File: sdk/spring/azure-spring-cloud-stream-binder-eventhubs/src/main/java/com/azure/spring/eventhub/stream/binder/EventHubMessageChannelBinder.java [hunk 2]
```diff
@@ -56,12 +56,12 @@ protected MessageHandler createProducerMessageHandler(ProducerDestination destin
         handler.setSendTimeout(producerProperties.getExtension().getSendTimeout());
         handler.setSendFailureChannel(errorChannel);
         if (producerProperties.isPartitioned()) {
-            handler.setPartitionKeyExpressionString(
-                    "'partitionKey-' + headers['" + BinderHeaders.PARTITION_HEADER + "']");
+            if (producerProperties.getPartitionKeyExpression() != null) {
+                handler.setPartitionKeyExpression(producerProperties.getPartitionKeyExpression());
+            }
```
Comment (saragluna) [inline sdk/spring/azure-spring-cloud-stream-binder-eventhubs/src/main/java/com/azure/spring/eventhub/stream/binder/EventHubMessageChannelBinder.java:61]: why this?

File: sdk/spring/azure-spring-integration-core/src/main/java/com/azure/spring/integration/core/api/reactor/DefaultMessageHandler.java [hunk 3]
```diff
@@ -161,17 +161,23 @@ private String toDestination(Message<?> message) {
...
-        String partitionKey = message.getHeaders().get(AzureHeaders.PARTITION_KEY, String.class);
-        if (!StringUtils.hasText(partitionKey) && this.partitionKeyExpression != null) {
-            partitionKey = this.partitionKeyExpression.getValue(this.evaluationContext, message, String.class);
+        // Priority setting partitionId
+        String partitionId = message.getHeaders().get(AzureHeaders.PARTITION_ID, String.class);
+        if (!StringUtils.hasText(partitionId) && message.getHeaders().containsKey(AzureHeaders.PARTITION_OVERRIDE)) {
+            partitionId = String.valueOf(message.getHeaders().get(AzureHeaders.PARTITION_OVERRIDE));
+        } else if (!StringUtils.hasText(partitionId) && message.getHeaders().containsKey(AzureHeaders.PARTITION_HEADER)) {
+            partitionId = String.valueOf(message.getHeaders().get(AzureHeaders.PARTITION_HEADER, Integer.class));
         }
```
Comment (saragluna) [inline sdk/spring/azure-spring-integration-core/src/main/java/com/azure/spring/integration/core/api/reactor/DefaultMessageHandler.java:170]: please try to refactor the code to make it easier to read

**Review by saragluna — CHANGES_REQUESTED** [review]

File: sdk/spring/azure-spring-integration-core/src/main/java/com/azure/spring/integration/core/api/reactor/DefaultMessageHandler.java (hunk 3 above)
Comment (moarychan) [reply sdk/spring/azure-spring-integration-core/src/main/java/com/azure/spring/integration/core/api/reactor/DefaultMessageHandler.java]: Refactored.

File: sdk/spring/azure-spring-cloud-stream-binder-eventhubs/src/main/java/com/azure/spring/eventhub/stream/binder/EventHubMessageChannelBinder.java (hunk 2 above)
Comment (moarychan) [reply sdk/spring/azure-spring-cloud-stream-binder-eventhubs/src/main/java/com/azure/spring/eventhub/stream/binder/EventHubMessageChannelBinder.java:61]: This is to support the configuration of the key partition expression provided by spring cloud stream.

File: sdk/spring/azure-spring-cloud-stream-binder-eventhubs/src/main/java/com/azure/spring/eventhub/stream/binder/EventHubMessageChannelBinder.java [hunk 4]
```diff
@@ -56,9 +56,7 @@ protected MessageHandler createProducerMessageHandler(ProducerDestination destin
         handler.setSendTimeout(producerProperties.getExtension().getSendTimeout());
         handler.setSendFailureChannel(errorChannel);
         if (producerProperties.isPartitioned()) {
-            if (producerProperties.getPartitionKeyExpression() != null) {
-                handler.setPartitionKeyExpression(producerProperties.getPartitionKeyExpression());
-            }
+            handler.setPartitionIdExpressionString("headers['" + BinderHeaders.PARTITION_HEADER + "']");
```
Comment (saragluna) [inline sdk/spring/azure-spring-cloud-stream-binder-eventhubs/src/main/java/com/azure/spring/eventhub/stream/binder/EventHubMessageChannelBinder.java:59]: I think we could remove the two properties: partitionIdExpressionString and partitionKeyExpressionString, and directly parse the string to an expression here, and then pass it to the handler.

File: sdk/spring/azure-spring-integration-core/src/main/java/com/azure/spring/integration/core/api/reactor/DefaultMessageHandler.java [hunk 5]
```diff
@@ -164,12 +169,9 @@ private PartitionSupplier toPartitionSupplier(Message<?> message) {
...
         MessageHeaders headers = message.getHeaders();
         // Priority setting partitionId
-        String partitionId = headers.get(AzureHeaders.PARTITION_ID, String.class);
-        if (!StringUtils.hasText(partitionId) && headers.containsKey(AzureHeaders.PARTITION_OVERRIDE)) {
-            partitionId = String.valueOf(headers.get(AzureHeaders.PARTITION_OVERRIDE));
-        } else if (!StringUtils.hasText(partitionId) && headers.containsKey(AzureHeaders.PARTITION_HEADER)) {
-            partitionId = String.valueOf(headers.get(AzureHeaders.PARTITION_HEADER, Integer.class));
-        }
+        String partitionId = this.partitionIdExpression != null
+            ? String.valueOf(this.partitionIdExpression.getValue(this.evaluationContext, message, Integer.class))
```
Comment (saragluna) [inline sdk/spring/azure-spring-integration-core/src/main/java/com/azure/spring/integration/core/api/reactor/DefaultMessageHandler.java:173]: can we directly get a string value from the expression?

File: sdk/spring/azure-spring-integration-core/src/main/java/com/azure/spring/integration/core/api/reactor/DefaultMessageHandler.java [hunk 6]
```diff
@@ -148,8 +149,12 @@ public void setPartitionKeyExpression(Expression partitionKeyExpression) {
...
-        setPartitionKeyExpression(EXPRESSION_PARSER.parseExpression(partitionKeyExpression));
+    public void setPartitionIdExpression(Expression partitionIdExpression) {
+        this.partitionIdExpression = partitionIdExpression;
+    }
+
+    public void setPartitionIdExpressionString(String partitionIdExpressionString) {
+        setPartitionIdExpression(EXPRESSION_PARSER.parseExpression(partitionIdExpressionString));
     }
 
     private String toDestination(Message<?> message) {
```
Comment (saragluna) [inline sdk/spring/azure-spring-integration-core/src/main/java/com/azure/spring/integration/core/api/reactor/DefaultMessageHandler.java:160]: do we have enough test cases for this?

**Review by saragluna — CHANGES_REQUESTED** [review]

**Review by saragluna — COMMENTED** [review]
Still, we need to add tests.

File: sdk/spring/azure-spring-integration-eventhubs/src/test/java/com/azure/spring/integration/eventhub/outbound/EventHubMessageHandlerTest.java [hunk 7]
```diff
@@ -33,4 +57,72 @@ public void setUp() {
         this.handler = new DefaultMessageHandler(this.destination, this.sendOperation);
     }
 
+    private void initProperties() {
```
Comment (saragluna) [inline sdk/spring/azure-spring-integration-eventhubs/src/test/java/com/azure/spring/integration/eventhub/outbound/EventHubMessageHandlerTest.java:60]: move this into setup function will make it call before each test case

File: sdk/spring/azure-spring-integration-eventhubs/src/test/java/com/azure/spring/integration/eventhub/outbound/EventHubMessageHandlerTest.java [hunk 8]
```diff
@@ -33,4 +57,72 @@ public void setUp() {
...
+            DefaultMessageHandler.class, TO_PARTITION_SUPPLIER_METHOD_NAME, message);
+        assertThat(partitionSupplier.getPartitionId()).isEqualTo("1");
+        assertThat(partitionSupplier.getPartitionKey()).isNull();
+
+        message = new GenericMessage<>(payloadBytes, ImmutableMap.of(AzureHeaders.PARTITION_ID, "2"));
+        this.handler.setPartitionIdExpression(partitionIdExpression);
+        partitionSupplier = ReflectionTestUtils.invokeMethod(this.handler,
+            DefaultMessageHandler.class, TO_PARTITION_SUPPLIER_METHOD_NAME, message);
+        assertThat(partitionSupplier.getPartitionId()).isEqualTo("2");
+        assertThat(partitionSupplier.getPartitionKey()).isNull();
```
Comment (saragluna) [inline sdk/spring/azure-spring-integration-eventhubs/src/test/java/com/azure/spring/integration/eventhub/outbound/EventHubMessageHandlerTest.java:84]: This code snippet is similar to above, perhaps we could remove this.

File: sdk/spring/azure-spring-integration-eventhubs/src/test/java/com/azure/spring/integration/eventhub/outbound/EventHubMessageHandlerTest.java [hunk 9]
```diff
@@ -33,4 +57,72 @@ public void setUp() {
...
+    public void testToPartitionSupplierReturnPartitionIdFromPartitionIdExpression() {
+        initProperties();
+
+        this.handler.setPartitionIdExpression(partitionIdExpression);
+        Message<?> message = new GenericMessage<>(payloadBytes, ImmutableMap.of("scst_partition", 3));
+        PartitionSupplier partitionSupplier = ReflectionTestUtils.invokeMethod(this.handler,
+            DefaultMessageHandler.class, TO_PARTITION_SUPPLIER_METHOD_NAME, message);
+        assertThat(partitionSupplier.getPartitionId()).isEqualTo("3");
+        assertThat(partitionSupplier.getPartitionKey()).isNull();
+    }
```
Comment (saragluna) [inline sdk/spring/azure-spring-integration-eventhubs/src/test/java/com/azure/spring/integration/eventhub/outbound/EventHubMessageHandlerTest.java:97]: add a case when we provide both AzureHeaders.PARTITION_ID and scst_partition

File: sdk/spring/azure-spring-integration-eventhubs/src/test/java/com/azure/spring/integration/eventhub/outbound/EventHubMessageHandlerTest.java [hunk 10]
```diff
@@ -33,4 +57,72 @@ public void setUp() {
...
+    public void testToPartitionSupplierReturnPartitionKeyFromHeader() {
+        initProperties();
+
+        Message<?> message = new GenericMessage<>(payloadBytes, ImmutableMap.of(AzureHeaders.PARTITION_KEY, "key1"));
+        PartitionSupplier partitionSupplier = ReflectionTestUtils.invokeMethod(this.handler,
+            DefaultMessageHandler.class, TO_PARTITION_SUPPLIER_METHOD_NAME, message);
+        assertThat(partitionSupplier.getPartitionId()).isNull();
+        assertThat(partitionSupplier.getPartitionKey()).isEqualTo("key1");
+
+        message = new GenericMessage<>(payloadBytes, ImmutableMap.of(AzureHeaders.PARTITION_KEY, "key2"));
```
Comment (saragluna) [inline sdk/spring/azure-spring-integration-eventhubs/src/test/java/com/azure/spring/integration/eventhub/outbound/EventHubMessageHandlerTest.java:109]: same here, why do we need this twice?

File: sdk/spring/azure-spring-integration-eventhubs/src/test/java/com/azure/spring/integration/eventhub/outbound/EventHubMessageHandlerTest.java [hunk 11]
```diff
@@ -33,4 +57,72 @@ public void setUp() {
...
+    public void testToPartitionSupplierReturnPartitionKeyFromPartitionKeyExpression() {
+        initProperties();
+
+        this.handler.setPartitionKeyExpression(partitionKeyExpression);
+        Message<?> message = new GenericMessage<>(payloadBytes);
+        PartitionSupplier partitionSupplier = ReflectionTestUtils.invokeMethod(this.handler,
+            DefaultMessageHandler.class, TO_PARTITION_SUPPLIER_METHOD_NAME, message);
+        assertThat(partitionSupplier.getPartitionId()).isNull();
+        assertThat(partitionSupplier.getPartitionKey()).isEqualTo(String.valueOf(payloadBytes.hashCode()));
+    }
```
Comment (saragluna) [inline sdk/spring/azure-spring-integration-eventhubs/src/test/java/com/azure/spring/integration/eventhub/outbound/EventHubMessageHandlerTest.java:127]: same here, add a case for both partitionkey header and partitionkeyExpression.

**Review by saragluna — CHANGES_REQUESTED** [review]

File: sdk/spring/azure-spring-integration-eventhubs/src/test/java/com/azure/spring/integration/eventhub/outbound/EventHubMessageHandlerTest.java (hunk 7 above)
Comment (moarychan) [reply sdk/spring/azure-spring-integration-eventhubs/src/test/java/com/azure/spring/integration/eventhub/outbound/EventHubMessageHandlerTest.java]: Comments added for this method.

File: sdk/spring/azure-spring-integration-eventhubs/src/test/java/com/azure/spring/integration/eventhub/outbound/EventHubMessageHandlerTest.java (hunk 10 above)
Comment (moarychan) [reply sdk/spring/azure-spring-integration-eventhubs/src/test/java/com/azure/spring/integration/eventhub/outbound/EventHubMessageHandlerTest.java]: Comments added for the reason.

**Review by saragluna — APPROVED** [review]
LGTM.

**Pull request merged**